Add recipe for ob-ess-julia

frederic-santos opened the pull request:
### Brief summary of what the package does

`ob-ess-julia` adds a lightweight support for Julia language with Org babel.

It is based on the orphaned package [`ob-julia`](https://github.com/gjkerns/ob-julia) by G. J. Kerns, and uses Emacs Speaks Statistics in the background.

An [example org file](https://github.com/frederic-santos/ob-ess-julia/blob/master/examples-ob-ess-julia.org) is available to test it (this obviously needs Julia to be installed).

### Direct link to the package repository

https://github.com/frederic-santos/ob-ess-julia

### Your association with the package

I am the author and maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### Notes
#### Byte compile warnings

There are a few warnings about unused lisp variables, but they correspond to features still to be further implemented.

#### Package-lint warnings

There are numerous instances where `package-lint` complains about function names and their prefixes. They cannot be changed, since their name is imposed by `ob-core.el`.

riscy commented:
Thanks for this - quick first pass.

### ob-ess-julia.el ###

byte-compile (using Emacs 27.1):
```
ob-ess-julia.el:28:1:Error: Cannot open load file: No such file or directory, julia-mode
```
I think this needs julia-mode in its Package-Requires.

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200906.512):
```
24 issues found:
15:38: warning: Use a non-snapshot version number for dependency on "ess" if possible.
43:0: error: "org-babel-ess-julia-external-command" doesn't start with package's prefix "ob-ess-julia".
54:0: error: "run-julia-and-select-buffer" doesn't start with package's prefix "ob-ess-julia".
61:0: error: "org-babel-ess-julia-eoe-indicator" doesn't start with package's prefix "ob-ess-julia".
63:0: error: "org-babel-ess-julia-eoe-output" doesn't start with package's prefix "ob-ess-julia".
78:0: error: "org-babel-ess-julia-default-session" doesn't start with package's prefix "ob-ess-julia".
81:0: error: "org-babel-header-args:ess-julia" doesn't start with package's prefix "ob-ess-julia".
81:0: error: `org-babel-header-args:ess-julia' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
98:0: error: "org-babel-ess-julia-initiate-session" doesn't start with package's prefix "ob-ess-julia".
132:0: error: "org-babel-ess-julia-associate-session" doesn't start with package's prefix "ob-ess-julia".
178:0: error: "org-babel-edit-prep:ess-julia" doesn't start with package's prefix "ob-ess-julia".
178:0: error: `org-babel-edit-prep:ess-julia' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
192:0: error: "org-babel-ess-julia-evaluate" doesn't start with package's prefix "ob-ess-julia".
204:0: error: "org-babel-expand-body:ess-julia" doesn't start with package's prefix "ob-ess-julia".
204:0: error: `org-babel-expand-body:ess-julia' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
227:0: error: "org-babel-ess-julia-write-object-command" doesn't start with package's prefix "ob-ess-julia".
235:0: error: "org-babel-ess-julia-evaluate-external-process" doesn't start with package's prefix "ob-ess-julia".
264:0: error: "org-babel-ess-julia-evaluate-session" doesn't start with package's prefix "ob-ess-julia".
373:0: error: "org-babel-ess-julia-process-value-result" doesn't start with package's prefix "ob-ess-julia".
383:0: error: "org-babel-ess-julia-graphical-output-file" doesn't start with package's prefix "ob-ess-julia".
389:0: error: "org-babel-load-session:ess-julia" doesn't start with package's prefix "ob-ess-julia".
389:0: error: `org-babel-load-session:ess-julia' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
398:0: error: "org-babel-ess-julia-quote-csv-field" doesn't start with package's prefix "ob-ess-julia".
408:0: error: "org-babel-ess-julia-assign-elisp" doesn't start with package's prefix "ob-ess-julia".
```
I'm aware of the org-babel convention, so this is okay.  You can use a [non-snapshot version of ess](https://stable.melpa.org/#/ess) however, and the `run-julia...` function should be prefixed.

Other possible lints:
- ob-ess-julia.el#L342: It's safer to sharp-quote function names; use `#'`
- ob-ess-julia.el#L12: Prefer `;; Version` over `;; Package-Version` (MELPA automatically adds `Package-Version`)

### Loadability ###

Verifying ability to #'load each file:
```
Loading ob-ess-julia.el
ob-ess-julia.el:Error: Emacs 27.1 errored during load
Done.
```

### Package ###

- Prefer the default recipe if possible.
- GitHub API found `Other`
  - Try to use a standard format for your license file - it looks like [CeCill should be supported](https://github.com/licensee/licensee/search?q=cecill)
    See: https://github.com/licensee/licensee
- I would recommend an [SPDX-License-Identifier](https://spdx.dev/ids/) to ob-ess-julia.el -- slightly more formal/detectable

frederic-santos commented:
Thanks for your review!

### Fixed
> I think this needs julia-mode in its Package-Requires. [...]
> the `run-julia...` function should be prefixed.
>  ob-ess-julia.el#L342: It's safer to sharp-quote function names; use #'
>  ob-ess-julia.el#L12: Prefer ;; Version over ;; Package-Version (MELPA automatically adds Package-Version)

These remarks have all been addressed.

> Try to use a standard format for your license file

I had quite a hard time in understanding what was the problem, but it just seems that only the *french* version of the license file is correctly recognized by Github and Licensee. The english version is unknown. I replaced the english version by the french one and everything is okay now (although maybe less explicit for non-french speakers).

> I would recommend an SPDX-License-Identifier to ob-ess-julia.el

I added this identifier in the two files of the package.

### Not sure if this is okay
> Prefer the default recipe if possible.

Actually I really need the `.jl` file in this package, thus I need a custom recipe.

> You can use a non-snapshot version of ess

The latest stable version of `julia-mode` has been specified. However, the latest stable version of ESS (18.10.2) is quite old [and does not work](https://github.com/frederic-santos/ob-ess-julia/issues/1#issuecomment-713869500) with `ob-ess-julia`. A newer (snapshot) version is needed.

> `ob-ess-julia.el:Error: Emacs 27.1 errored during load`

Don't really know why this error was triggered. I simply removed the version dependency.

Thanks again!

riscy commented:
> I had quite a hard time in understanding what was the problem, but it just seems that only the french version of the license file is correctly recognized by Github and Licensee.

Wow!  Thank you -- hopefully that wasn't too burdensome.

> Don't really know why this error was triggered. I simply removed the version dependency.

This was just because of the Julia requirement (probably I didn't need to include it).  It looks like the linters will remain quite noisy due to org-babel conventions, but this looks ready.

And, welcome to MELPA!
